robots/flakefinder/report.go

```diff
@@ -348,6 +349,67 @@ func Report(results []*Result, reportOutputWriter *storage.Writer) error {
 	return nil
 }
 
+// SortTestsByRelevance sorts given tests according to the severity from the test data, where tests with a higher
+// severity have a smaller index in the slice than tests with a lower severity.
+// The returned slice does not contain
+// duplicates, thus if a test has data with several severities the highest one is picked, leading to an earlier
+// encounter in the slice.
+func SortTestsByRelevance(data map[string]map[string]*Details, tests []string) (testsSortedByRelevance []string) {
+	foundTests := map[string]bool{}
+	for _, test := range tests {
+		foundTests[test] = false
+	}
+
+	// Group all tests by severity, ignoring duplicates for the moment, but keeping a record of tests
+	// that have not been found
+	flakinessToTestNames := map[string][]string{}
+	for test, jobsToDetails := range data {
+		for _, details := range jobsToDetails {
+			if _, exists := flakinessToTestNames[details.Severity]; !exists {
+				flakinessToTestNames[details.Severity] = []string{}
+			}
+			flakinessToTestNames[details.Severity] = append(flakinessToTestNames[details.Severity], test)
+
+			foundTests[test] = true
+		}
+	}
+
+	// Build up the initial sorted result (with duplicates)
+	initialTestsSortedByRelevance := append(flakinessToTestNames[HeavilyFlaky])
+	initialTestsSortedByRelevance = append(initialTestsSortedByRelevance, flakinessToTestNames[MostlyFlaky]...)
+	initialTestsSortedByRelevance = append(initialTestsSortedByRelevance, flakinessToTestNames[ModeratelyFlaky]...)
+	initialTestsSortedByRelevance = append(initialTestsSortedByRelevance, flakinessToTestNames[MildlyFlaky]...)
+	initialTestsSortedByRelevance = append(initialTestsSortedByRelevance, flakinessToTestNames[Fine]...)
+	initialTestsSortedByRelevance = append(initialTestsSortedByRelevance, flakinessToTestNames[Unimportant]...)
+
+	// Append all tests that have not been found in the data
+	for _, test := range tests {
+		if !foundTests[test] {
+			initialTestsSortedByRelevance = append(initialTestsSortedByRelevance, test)
+		}
+	}
+
+	// Now kill the duplicates by keeping a map of whether the test was encountered before
```

Comment:
This could probably be done while appending the tests in the first pass, but should not be a problem since the small cardinality of the data set.